### PR TITLE
Add fused glu grad

### PR DESCRIPTION
**Pull request by zobinHuang**

### CODE REVIEW

**Comment by zobinHuang** [comment]
# 前向 & 反向整理逻辑

## 情况 I：参数矩阵合并
fused_glu

## 情况 II：参数矩阵分离
fused_glu_split

**Comment by zobinHuang** [comment]
**记录：float32 下 `fast_gelu` 激活函数偶尔会发生精度损失问题**

======================================================================
FAIL: test_gather (__main__.TestFusedGlu)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_fused_glu.py", line 216, in test_gather
    arg[0](test_case, *arg[1:])
  File "./test_fused_glu.py", line 134, in _test_fused_glu
    compare_result(test_case, fused_w_grad, w_grad)
  File "./test_fused_glu.py", line 106, in compare_result
    f"\na\n{a.numpy()}\n{'-' * 80}\nb:\n{b.numpy()}\n{'*' * 80}\ndiff:\n{a.numpy() - b.numpy()}",
AssertionError: False is not true : 
a
[[ 2.2851196e-04  2.4142621e-04  2.8132717e-04 ... -2.4457290e-04
   1.0508957e-05  5.0922430e-05]
 [ 8.1784419e-05  3.2076027e-04  9.8911274e-05 ...  8.2907740e-05
  -1.2322213e-04  1.7557791e-04]
 [ 2.5440060e-04  5.8088721e-05  3.5915125e-04 ... -1.4365767e-04
  -3.1520874e-05  1.3237423e-04]
 ...
 [-2.3484584e-04 -4.9004430e-04  4.6607910e-04 ... -1.4304393e-04
  -1.4439087e-04  4.4898965e-04]
 [-2.2394922e-04 -1.8557499e-04 -1.0166231e-04 ...  1.8585197e-05
  -2.2483224e-04 -2.2874537e-04]
 [ 4.9691898e-04 -3.8029586e-05 -3.5308243e-04 ...  4.0103751e-06
  -6.5397493e-05  3.7856083e-04]]
--------------------------------------------------------------------------------
b:
[[ 2.2851196e-04  2.4142621e-04  2.8132717e-04 ... -2.4457290e-04
   1.0508957e-05  5.0922430e-05]
 [ 8.1784419e-05  3.2076027e-04  9.8911274e-05 ...  8.2907740e-05
  -1.2322213e-04  1.7557791e-04]
 [ 2.5440060e-04  5.8088721e-05  3.5915125e-04 ... -1.4365767e-04
  -3.1520874e-05  1.3237423e-04]
 ...
 [-2.3484584e-04 -4.9004430e-04  4.6607910e-04 ... -1.4304393e-04
  -1.4439087e-04  4.4898965e-04]
 [-2.2394922e-04 -1.8557499e-04 -1.0166231e-04 ...  1.8585197e-05
  -2.2483224e-04 -2.2874537e-04]
 [ 4.9691898e-04 -3.8029586e-05 -3.5308243e-04 ...  4.0103751e-06
  -6.5397493e-05  3.7856083e-04]]
********************************************************************************
diff:
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]

----------------------------------------------------------------------

File: oneflow/user/ops/fused_glu_without_linear_grad_op.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+          .Split(user_op::OpArg("matmul_vx", 0), i)
+          .Split(ctx->outputs(), i)
+          .Build();
+    } else {
+      ctx->NewBuilder()
+          .Split(user_op::OpArg("dy", 0), i)
+          .Split(user_op::OpArg("matmul_wx", 0), i)
+          .Split(ctx->outputs(), i)
+          .Build();
+    }
```
Comment (yz-chen18) [inline oneflow/user/ops/fused_glu_without_linear_grad_op.cpp:43]: 反向传播时，对m维进行split，X的m维应该也需要进行split

File: oneflow/core/ep/common/primitive/binary_functor.h [hunk 2]
```diff
@@ -359,6 +366,17 @@ struct BinaryFunctor<device, BinaryOp::kReluBackwardWithDyY, Src, Dst> {
...
 };
 
+template<DeviceType device, typename Src, typename Dst>
+struct BinaryFunctor<device, BinaryOp::kReluBackwardWithDyX, Src, Dst> {
+  OF_DEVICE_FUNC BinaryFunctor(Scalar attr0, Scalar attr1) {}
+
+  OF_DEVICE_FUNC Dst operator()(Src dy, Src x) const {
+    const Src zero_val = static_cast<Src>(0.0);
+    const Src y = x <= zero_val ? zero_val : x;
+    return static_cast<Dst>((y <= static_cast<Src>(0.0)) ? static_cast<Src>(0.0) : dy);
```
Comment (yz-chen18) [inline oneflow/core/ep/common/primitive/binary_functor.h:376]: 这里的y是不是一直非负了啊
```suggestion
    return static_cast<Dst>((x <= static_cast<Src>(0.0)) ? static_cast<Src>(0.0) : dy);
```

File: oneflow/user/kernels/fused_glu_without_linear_grad_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    const uintptr_t matmul_wx_ptr = reinterpret_cast<uintptr_t>(matmul_wx);
+    const uintptr_t matmul_vx_ptr = reinterpret_cast<uintptr_t>(matmul_vx);
+    const uintptr_t d_matmul_wx_ptr = reinterpret_cast<uintptr_t>(d_matmul_wx);
+    const uintptr_t d_matmul_vx_ptr = reinterpret_cast<uintptr_t>(d_matmul_vx);
+    return (/* memory address alignment */
+            dy_ptr % alignment == 0 && matmul_vx_ptr % alignment == 0
+            && matmul_wx_ptr % alignment == 0 && d_matmul_wx_ptr % alignment == 0
+            && d_matmul_vx_ptr % alignment == 0
+            /* #element per row alignment */
+            && n % (alignment / sizeof(T)) == 0);
```
Comment (yz-chen18) [inline oneflow/user/kernels/fused_glu_without_linear_grad_kernel.cu:168]: 如果alignment / sizeof(T) == 0，则会发生n % 0的未定义行为，或许在168行前对alignment / sizeof(T)进行检查

**Comment by zobinHuang** [comment]
## 初步性能测试

image

File: oneflow/user/kernels/fused_glu_kernel.cu [hunk 4]
```diff
@@ -25,12 +25,17 @@ limitations under the License.
 #include "oneflow/core/ep/cuda/primitive/unary_functor.cuh"
 #include "oneflow/core/kernel/util/cuda_half_util.h"
 #include "oneflow/core/kernel/cuda_graph_support.h"
+#include "oneflow/user/kernels/cublas_fused_mlp_util.cuh"
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_glu_kernel.cu:28]: 这里等 https://github.com/Oneflow-Inc/oneflow/pull/9724 处理完

File: oneflow/user/kernels/fused_glu_kernel.cu [hunk 5]
```diff
@@ -25,12 +25,17 @@ limitations under the License.
...
 
 #if CUDA_VERSION >= 11000
 #include <cuda_bf16.h>
 #endif  // CUDA_VERSION >= 11000
 #include "oneflow/core/device/cuda_pseudo_bfloat16.h"
 
+#if CUDA_VERSION >= 10020
+
+#if !defined(__clang__)
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_glu_kernel.cu:37]: WITH_CUTLASS 已经排除 __clang__ 了

File: oneflow/user/kernels/fused_glu_kernel.cu [hunk 6]
```diff
@@ -634,3 +700,5 @@ REGISTER_GPU_FUSED_GLU_KERNEL(nv_bfloat16)
 #endif
 
 }  // namespace oneflow
+
+#endif  // CUDA_VERSION >= 10020
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_glu_kernel.cu:704]: 结尾的换行符

File: oneflow/user/kernels/fused_glu_kernel.cu [hunk 7]
```diff
@@ -584,35 +555,130 @@ class GpuFusedGluKernel final : public user_op::OpKernel, public user_op::CudaGr
...
+          /*A*/ input_tensor_w->dptr(),
+          /*Adesc*/ fused_glu_cache->cublas_a_desc,
+          /*B*/ input_tensor_x->dptr(),
+          /*Bdesc*/ fused_glu_cache->cublas_b_desc,
+          /*beta*/ &sp_beta,
+          /*C*/ input_tensor_b->dptr(),
+          /*Cdesc*/ fused_glu_cache->cublas_c_desc,
+          /*D*/ out_tensor_matmul_wx->mut_dptr(),
+          /*Ddesc*/ fused_glu_cache->cublas_c_desc,
+          /*algo*/ nullptr,
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_glu_kernel.cu:601]: 这里看一下 fused_matmul_bias的最新实现，会有一点不一样

File: oneflow/user/ops/fused_glu_without_linear_grad_op.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+       ++i) {
+    if (is_split_mode) {
+      ctx->NewBuilder()
+          .Split(user_op::OpArg("dy", 0), i)
+          .Split(user_op::OpArg("matmul_wx", 0), i)
+          .Split(user_op::OpArg("matmul_vx", 0), i)
+          .Split(ctx->outputs(), i)
+          .Build();
+    } else {
+      ctx->NewBuilder()
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_glu_without_linear_grad_op.cpp:36]: 最后一维是不是不能split的

File: oneflow/user/ops/fused_glu_without_linear_grad_op.cpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  bool is_split_mode = false;
+  if (ctx->has_input("matmul_vx", 0)) { is_split_mode = true; }
+
+  // obtain dimensions of dy and matmul_wx
+  size_t dy_num_axes = dy_shape.NumAxes();
+  size_t matmul_wx_num_axes = matmul_wx_shape.NumAxes();
+
+  // check dimensions of dy and matmul_wx
+  CHECK_GE_OR_RETURN(dy_num_axes, 2)
+      << "number of axes of \'dy\' should have be greater than 1, yet get " << dy_num_axes;
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_glu_without_linear_grad_op.cpp:63]: 这种地方最好代码和注释统一，可以考虑将上面替换为 CHECK_GT_OR_RETURN， 或者改下面的出错信息

File: oneflow/user/kernels/fused_glu_kernel.cu [hunk 10]
```diff
@@ -25,12 +25,15 @@ limitations under the License.
...
 #include "oneflow/core/kernel/util/cuda_half_util.h"
 #include "oneflow/core/kernel/cuda_graph_support.h"
+#include "oneflow/user/kernels/cublas_fused_mlp_util.cuh"
 
 #if CUDA_VERSION >= 11000
 #include <cuda_bf16.h>
 #endif  // CUDA_VERSION >= 11000
 #include "oneflow/core/device/cuda_pseudo_bfloat16.h"
 
+#if CUDA_VERSION >= 10020
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_glu_kernel.cu:35]:
```suggestion
#if CUDA_VERSION >= 11020
```

**Review by liujuncheng — APPROVED** [review]

**Review by yz-chen18 — APPROVED** [review]

**Pull request merged**